### PR TITLE
(feat.): Multi avatar support and simplified APIs for ```ActerAvatar```.

### PR DESCRIPTION
**Pull request by gtalha07**
The PR now brings the multi parent badges support for default avatars along with few adjustments and simplifications:
- ```ActerAvatar``` widget now only supports ```AvatarOptions``` which can be regarded as simpler outer wrapper for ```AvatarInfo``` along with other necessary informations: size, tooltip etc. This approach allows a lot of reduction in code duplication, and easier understanding in the flow of elements. 
- ```AvatarOptions``` exists in 3 variants:
 1. `AvatarOptions` as ```DefaultAvatarOptions```. This is similar to its precessor modes `DisplayMode.Space`, `DisplayMode.GroupChat`. These modes are synonymous in terms of their implementation, so they've been merged.
 2. `AvatarOptions` as `AvatarOptions.DM()`. This is similar to its precessor modes `DisplayMode.DM`.
 3. `AvatarOptions` as `AvatarOptions.GroupDM()`. This is similar to its precessor modes `DisplayMode.GroupDM`.
 
 This also includes multi badges support for default `ActerAvatar` mode .i.e. DisplayMode.Space/GroupChat in legacy terms as show in below screenshot. 
Screenshot 2024-05-21 at 9 40 20 PM

Pre-merge essentials:

- [x] Update Documentation.
- [x] Revise the version with breaking changes announcement.

### CODE REVIEW

File: lib/src/acter_avatar.dart [hunk 1]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-            borderRadius: BorderRadius.circular(6.0),
-            image: DecorationImage(
-              fit: BoxFit.cover,
-              image: avatar,
-              onError: avatarError,
+        // render for DM
+        DM: (avatarInfo, size) => CircleAvatar(
+              foregroundImage: avatar,
+              onForegroundImageError: avatarError,
+              radius: size ?? 24,
```
Comment (gnunicorn) [inline lib/src/acter_avatar.dart:177]: the diffs here make it hard to follow what has actually changed. I glanced over these and tried to find difference but failed to do so. If there is anything substantively different I should be looking it, please point it out and I will.

File: lib/src/acter_avatar.dart [hunk 2]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-                              decoration: const BoxDecoration(
-                                color: Colors.green,
-                                shape: BoxShape.circle,
-                              ),
-                              child: Text(
-                                '+${widget.avatarsInfo!.length - 1}',
-                                style: const TextStyle(fontSize: 8),
+        // render for Group DM
+        GroupDM: ((avatarInfo, groupAvatarsInfo, size, groupAvatarSize) {
+          return groupAvatarsInfo != null && groupAvatarsInfo.isNotEmpty
```
Comment (gnunicorn) [inline lib/src/acter_avatar.dart:181]: I wonder if we shouldn't move all these into separate functions to make it easier to keep the overview of what is happenening when ... so for every when have a corresponding function being called with the parameters given ...

File: lib/src/acter_avatar.dart [hunk 3]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
+    final size = badgesSize ?? 20;
+    final badges = options.parentBadges!;
+    double badgeOverflow = size / 5;
+    final badgesLength = badges.length;
+    final thresholdCount = badgesLength <= 3 ? badgesLength : 3;
+    List<Widget> children = [];
+
+    for (int i = 0; i < thresholdCount; i++) {
+      children.add(
+        ActerAvatar(options: AvatarOptions(badges[i], size: size)),
```
Comment (gnunicorn) [inline lib/src/acter_avatar.dart:247]: I prefer the briefer
```suggestion
    List<Widget> children = badges.getRange(0, thresholdCount).map((badge) => ActerAvatar(options: AvatarOptions(badge, size: size))).toList(();

```

**Review by gnunicorn — APPROVED** [review]
Not able to follow all diffs in details, as the order also changed and thus makes it hard to compare all details with what we had before. But I couldn't find any large or obvious problems. Marked a few smaller things though.

File: lib/src/acter_avatar.dart (hunk 1 above)
Comment (gtalha07) [reply lib/src/acter_avatar.dart:177]: Yeah, the code is a bit complex to follow. I'll point out areas where it is meaningful to get understanding of whole picture 👍🏻 !

File: lib/src/acter_avatar.dart (hunk 2 above)
Comment (gtalha07) [reply lib/src/acter_avatar.dart:181]: I think we could move these into sub widget methods to enhance code readability.

File: lib/src/acter_avatar.dart [hunk 4]
```diff
@@ -12,48 +12,23 @@ final log = Logger('ActerAvatar');
 /// image resizing (for memory purposes) has to be done beforehand, the image
 /// is shown as given.
 class ActerAvatar extends StatefulWidget {
-  /// Set the display mode for this Avatar.
-  final DisplayMode mode;
```
Comment (gtalha07) [inline lib/src/acter_avatar.dart:16]: Firstly, the most major change in this PR is to wrap all these into one configurable parameter `AvatarOptions` which  can be initialised in different ways depending upon type of avatar. I found out that this approach was reducing a lot of code lines when played in actual widget tree.

File: lib/src/acter_avatar.dart [hunk 5]
```diff
@@ -115,32 +92,18 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-            widget.avatarInfo.uniqueName ?? widget.avatarInfo.uniqueId;
-        if (widget.avatarInfo.displayName != null) {
-          message =
-              '${widget.avatarInfo.displayName} (${widget.avatarInfo.uniqueId})';
-        }
-        return Tooltip(message: message, child: child);
-      case TooltipStyle.None:
-        return child;
-    }
+    return tooltipMsgWithChild(avatarInfo, child);
```
Comment (gtalha07) [inline lib/src/acter_avatar.dart:98]: This has also been reduced to have one singular widget method, both for main avatar and parent badges tooltip display.

File: lib/src/acter_avatar.dart [hunk 6]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-            Container(
-              height: widget.size ?? 48,
-              width: widget.size ?? 48,
-              decoration: BoxDecoration(
-                borderRadius: BorderRadius.circular(6.0),
-                image: DecorationImage(
-                  fit: BoxFit.cover,
-                  image: avatar,
-                  onError: avatarError,
+    return widget.options.when(
```
Comment (gtalha07) [inline lib/src/acter_avatar.dart:150]: I couldn't find any valid and easier approach that vehemently checks dart object initialisation by which constructor type. ```freezed``` seems to provide all necessary boilerplate generated methods, essentially making it easier to check object initialisation by constructor type. Although I believe this can be simplified further by wrapping into sub-widget methods as per your feedback.

File: lib/src/acter_avatar.dart [hunk 7]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-    if (widget.badgeSize == null) {
-      // nothing. ignore
-      return SizedBox.shrink();
-    }
-    final badgeSize = widget.badgeSize ?? 20;
-    if (widget.avatarsInfo == null || widget.avatarsInfo!.isEmpty) {
-      return SizedBox(
-        height: badgeSize,
-        width: badgeSize,
+  Widget renderBadges(
```
Comment (gtalha07) [inline lib/src/acter_avatar.dart:233]: Logically the code is pretty much same overall except for this part, here the code has been modified to support rendering multiple parent spaces .i.e. badges in plugin terms.

File: lib/src/acter_avatar.dart [hunk 8]
```diff
@@ -162,245 +126,242 @@ class _ActerAvatar extends State<ActerAvatar> {
...
-      case DisplayMode.GroupDM:
-        return widget.avatarsInfo != null && widget.avatarsInfo!.isNotEmpty
-            ? Stack(
-                alignment: Alignment.bottomLeft,
-                clipBehavior: Clip.none,
-                children: [
-                  MultiAvatar(
-                    uniqueId: widget.avatarInfo.uniqueId,
-                    size: multiFallbackSize,
+    final avatarOptions = widget.options;
```
Comment (gtalha07) [inline lib/src/acter_avatar.dart:284]: The changes in methods only reflects switching avatar type fallback based on ```AvatarOptions``` type constructor instead of old way enum.

File: lib/src/models/avatar_options.dart [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+/// Helper class for defining avatar type.
+/// Currently supported modes are Space/Group Chat (default), DM, Group DM.
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'avatar_options.freezed.dart';
+
+@freezed
+abstract class AvatarOptions with _$AvatarOptions {
+  /// Default usage .i.e. Space/Group Chat. Group avatars will appear as secondary badges.
```
Comment (gtalha07) [inline lib/src/models/avatar_options.dart:11]: I think enum values could be reduced to have one default constructor which can have mode of Space/GroupChat avatar with its optional badges params. The other types can be initialised via their named constructors.

File: lib/src/acter_avatar.dart (hunk 1 above)
Comment (gnunicorn) [reply lib/src/acter_avatar.dart]: I understand the picture overall, thanks for the pointers. But what I meant was if there was any changes in the way that things are rendered compared to before ... I dunno changes of sizes or stacking order or anything like that ... that is hard to find out.
